### ISSUE TITLE
iOS 11 suddenly stops receiving FCM remote notifications, sometimes never receives them at all

### ISSUE THREAD

**Issue by xxtesaxx** [firebase/quickstart-ios#403]
I got a report from a user that his iOS 11 device is not updating automatically. I use a background remote notification to pull data when new data is available. in iOS 10 this just works fine and when I tested it on my iOS 11 iPhone 6 it also worked.

But then I noticed that after a few minutes the iPhone 6 stopped receiving remote notifications as well and since then didn't receive any new messages. Even if I re-install the app it now does not get any more messages. 

The only way to receive a message is to include the "notification" key in the FCM Message with at least a "title". To prevent notifications from showing to the user, I would need to include a title of "" ("notification" : { "title":""},...) 

What bothers me the most is that it just worked an hour ago and no it doesnt work anymore at all. I already tried downgrading the FirebaseInstanceID pod to 2.0.0 but that didnt work out.

Is there something I'm not aware of like a new info.plist entry for iOS 11 which has to be set? Why did it then work for a few minutes before now not working anymore at all? Is this a known bug and if so, is there a workaround?

**Comment by morganchen12** [comment]
Make sure your code is handling FCM token refresh events. Downgrading your InstanceID version shouldn't have any effect, so use the latest InstanceID version if you can.

Without any code or repro steps, we won't really be able to help you debug further. This is why we provide the issue template. Please fill it out if possible.

**Comment by DennisMostajo** [comment]
I have the same problem since the previous week, I am managing all the events to update and manage the FCM token, my pod includes:

    pod 'Firebase’, ’4.0.4’
    pod 'Firebase/Messaging', '4.0.4'
    pod 'Firebase/Core', '4.0.4'
    pod 'FirebaseInstanceID', '2.0.0'

And my AppDelegate has:

`
        
        func initializeFCM(_ application: UIApplication)
    {
        print("initializeFCM")
         //-------------------------------------------------------------------------//
        // iOS 10 support
        if #available(iOS 10, *) {
            let center = UNUserNotificationCenter.current()
            center.delegate = self
            center.requestAuthorization(options:[.badge, .alert, .sound]){ (granted, error) in }
            application.registerForRemoteNotifications()
        }
            // iOS 9 support
        else if #available(iOS 9, *) {
            application.registerUserNotificationSettings(UIUserNotificationSettings(types: [.badge, .sound, .alert], categories: nil))
            application.registerForRemoteNotifications()
        }
        else {  
            application.registerForRemoteNotifications(matching: [.badge, .sound, .alert])
        }
      //-------------------------------------------------------------------------//
        if application.isRegisteredForRemoteNotifications
        {
            debugPrint("---> is Registered For Remote Notifications")
        }
        else
        {
            debugPrint("---> is NOT Registered For Remote Notifications")
        }                                                             
        FirebaseOptions.defaultOptions()?.deepLinkURLScheme = "my app n_n' " // Firebase Dynamic Link
        FirebaseApp.configure()
        Messaging.messaging().delegate = self
        Messaging.messaging().shouldEstablishDirectChannel = true
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.tokenRefreshNotificaiton),
                                                         name: NSNotification.Name.InstanceIDTokenRefresh, object: nil)
    }
    
    @objc func tokenRefreshNotificaiton(_ notification: Foundation.Notification) {
        if let refreshedToken = InstanceID.instanceID().token() {
            StorageHelper.setCurrentUserFCMID(refreshedToken)
            debugPrint("InstanceID token: \(refreshedToken)")
        }
    }
    
    @available(iOS 10.0, *)
    func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void)
    {
        //Handle the notification ON APP
        debugPrint("*** willPresent notification")
        debugPrint("*** notification: \(notification.request.content.userInfo)")
        Messaging.messaging().appDidReceiveMessage(notification.request.content.userInfo)
        completionHandler([.sound,.alert,.badge])
    }
    
    @available(iOS 10.0, *)
    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void)
    {
        //Handle the notification ON BACKGROUND
        debugPrint("TAP NOTIFICATION")
        debugPrint("*** didReceive response Notification ")
        debugPrint("*** response: \(response.notification.request.content.userInfo)")
        Messaging.messaging().appDidReceiveMessage(response.notification.request.content.userInfo)
            if let gcmMessageType = GCMMessageType(rawValue: notificationType) {
                debugPrint("didReceiveLocalNotification")
            }
        completionHandler()
    }`

it's strange because it was working and it also does not arrive on Android devices

**Comment by morganchen12** [comment]
Looks like you're probably sending `content-available` notifications.

The system (iOS) will throttle these to conserve resources. Apple doesn't explicitly detail any throttling rules and they may change between OS versions. Non-silent notifications aren't throttled, since they must be displayed to the user. Background sync may be disabled entirely if the user enables low power mode on their device, and throttling may go away if the device receiving notifications is plugged into a power source.

If your notifications stop entirely after a while, it's possible your app isn't correctly handling FCM token refresh events.

**Comment by DennisMostajo** [comment]
Hi @morganchen12 
, thanks for answer, my code is: 
```
`//
//  AppDelegate.swift
//  lol
//
//  Created by Dennis Mostajo on, you know when n_n'
//  Copyright © 2017 Dennis Mostajo. All rights reserved.
//

import UIKit

import FirebaseAnalytics
import FirebaseMessaging
import FirebaseInstanceID
import FirebaseDynamicLinks

//-------------------------------------------------------------------------//
import UserNotifications // enable new way for notifications on iOS 10+
//-------------------------------------------------------------------------//

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate , UNUserNotificationCenterDelegate, MessagingDelegate// enable new way for notifications on iOS 10
{

    /// This method will be called whenever FCM receives a new, default FCM token for your
    /// Firebase project's Sender ID.
    /// You can send this token to your application server to send notifications to this device.
    public func messaging(_ messaging: Messaging, didRefreshRegistrationToken fcmToken: String)
    {
        debugPrint("--->messaging didRefreshRegistrationToken")
        debugPrint("--->fcmToken:\(fcmToken)")
    }
    
    
    /// This method is called on iOS 10 devices to handle data messages received via FCM through its
    /// direct channel (not via APNS). For iOS 9 and below, the FCM data message is delivered via the
    /// UIApplicationDelegate's -application:didReceiveRemoteNotification: method.
    @available(iOS 10.0, *)
    public func messaging(_ messaging: Messaging, didReceive remoteMessage: MessagingRemoteMessage)
    {
        debugPrint("--->messagin didReceive Remote Message")
        guard let data =
            try? JSONSerialization.data(withJSONObject: remoteMessage.appData, options: .prettyPrinted),
            let prettyPrinted = String(data: data, encoding: .utf8) else { return }
        print("Received direct channel message:\n\(prettyPrinted)")
        self.application(UIApplication.shared, didReceiveRemoteNotification: remoteMessage.appData)
        self.application(UIApplication.shared, didReceiveRemoteNotification: remoteMessage.appData, fetchCompletionHandler:{_ in debugPrint("Firebase delegate handle remote notification")})
        Messaging.messaging().appDidReceiveMessage(remoteMessage.appData)
    }
    
    public func application(received remoteMessage: MessagingRemoteMessage)
    {
        debugPrint("--->application didReceive Remote Message")
        guard let data =
            try? JSONSerialization.data(withJSONObject: remoteMessage.appData, options: .prettyPrinted),
            let prettyPrinted = String(data: data, encoding: .utf8) else { return }
        print("Received direct channel message:\n\(prettyPrinted)")
        self.application(UIApplication.shared, didReceiveRemoteNotification: remoteMessage.appData)
        self.application(UIApplication.shared, didReceiveRemoteNotification: remoteMessage.appData, fetchCompletionHandler:{_ in debugPrint("Firebase delegate handle remote notification")})
        Messaging.messaging().appDidReceiveMessage(remoteMessage.appData)
    }
    
    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        
        self.initializeFCM(application)
        let token = InstanceID.instanceID().token()
        debugPrint("GCM TOKEN = \(String(describing: token))")
        
        return true
    }
    
    func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error)
    {
        debugPrint("didFailToRegisterForRemoteNotificationsWithError: \(error)")
    }
    
    func initializeFCM(_ application: UIApplication)
    {
        print("initializeFCM")
        
        //-------------------------------------------------------------------------//
        // iOS 10 support
        if #available(iOS 10, *) {
            let center = UNUserNotificationCenter.current()
            center.delegate = self
            center.requestAuthorization(options:[.badge, .alert, .sound]){ (granted, error) in }
            application.registerForRemoteNotifications()
        }
            // iOS 9 support
        else if #available(iOS 9, *) {
            application.registerUserNotificationSettings(UIUserNotificationSettings(types: [.badge, .sound, .alert], categories: nil))
            application.registerForRemoteNotifications()
        }
        else {
            application.registerForRemoteNotifications(matching: [.badge, .sound, .alert])
        }
        //-------------------------------------------------------------------------//
        if application.isRegisteredForRemoteNotifications
        {
            debugPrint("---> is Registered For Remote Notifications")
        }
        else
        {
            debugPrint("---> is NOT Registered For Remote Notifications")
        }
        
        FirebaseOptions.defaultOptions()?.deepLinkURLScheme = "my app n_n' " // Firebase Dynamic Link
        FirebaseApp.configure()
        Messaging.messaging().delegate = self
        Messaging.messaging().shouldEstablishDirectChannel = true
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.tokenRefreshNotificaiton),
                                               name: NSNotification.Name.InstanceIDTokenRefresh, object: nil)
    }
    
    @objc func tokenRefreshNotificaiton(_ notification: Foundation.Notification) {
        if let refreshedToken = InstanceID.instanceID().token() {
            debugPrint("InstanceID token: \(refreshedToken)")
        }
    }
    
    @available(iOS 10.0, *)
    func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void)
    {
        //Handle the notification ON APP
        debugPrint("*** willPresent notification")
        debugPrint("*** notification: \(notification.request.content.userInfo)")
        Messaging.messaging().appDidReceiveMessage(notification.request.content.userInfo)
        completionHandler([.sound,.alert,.badge])
    }
    
    @available(iOS 10.0, *)
    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void)
    {
        //Handle the notification ON BACKGROUND
        debugPrint("TAP NOTIFICATION")
        debugPrint("*** didReceive response Notification ")
        debugPrint("*** response: \(response.notification.request.content.userInfo)")
        Messaging.messaging().appDidReceiveMessage(response.notification.request.content.userInfo)
        completionHandler()
    }
    
    
    private func application(application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: NSData)
    {
        debugPrint("didRegisterForRemoteNotificationsWithDeviceToken: NSDATA")
        
        let token = String(format: "%@", deviceToken as CVarArg)
        debugPrint("*** deviceToken: \(token)")
        Messaging.messaging().apnsToken = deviceToken as Data
        debugPrint("Firebase Token:",InstanceID.instanceID().token() as Any)
    }
    
    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data)
    {
        debugPrint("didRegisterForRemoteNotificationsWithDeviceToken: DATA")
        let token = String(format: "%@", deviceToken as CVarArg)
        debugPrint("*** deviceToken: \(token)")
        let deviceTokenString = deviceToken.reduce("", {$0 + String(format: "%02X", $1)})
        debugPrint("deviceTokenString: \(deviceTokenString)")
        Messaging.messaging().apnsToken = deviceToken
        debugPrint("Firebase Token:",InstanceID.instanceID().token() as Any)
    }
    //-------------------------------------------------------------------------//
    
    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable: Any]) {
        // If you are receiving a notification message while your app is in the background,
        // this callback will not be fired till the user taps on the notification launching the application.
        // TODO: Handle data of notification
        // Print message ID.
        
        Messaging.messaging().appDidReceiveMessage(userInfo)
        
        if let messageID = userInfo["gcm.message_id"] {
            debugPrint("Message ID: \(messageID)")
        }
        debugPrint("---> didReceiveRemoteNotification")
    }
    
    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable: Any],
                     fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
        // If you are receiving a notification message while your app is in the background,
        // this callback will not be fired till the user taps on the notification launching the application.
        // TODO: Handle data of notification
        Messaging.messaging().appDidReceiveMessage(userInfo)
        
        if let messageID = userInfo["gcm.message_id"] {
            debugPrint("Message ID: \(messageID)")
        }
        debugPrint("---> didReceiveRemoteNotification completionHandler")
        
        // Print full message.
        completionHandler(.newData)
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }
}`

```
I'm receiving the GCM token, but what I'm missing for handle or update it?
